parlai/scripts/convert_fairseq_to_parlai.py

```diff
@@ -0,0 +1,485 @@
+
+        :return state_dict:
+            return a state dict to load into ParlAI model.
+        """
+        # deal with embeddings
+        state = self.state
+        agent = self.agent
+        state_dict = state['model']
+        return_dict = OrderedDict()
+        for each_key in state_dict.keys():
+            mapped_key = each_key
+            if mapped_key == 'encoder.version' or mapped_key == 'decoder.version':
+                continue
+
+            # 1. replace if embedding
+            for emb in EMBEDDING_DICT_MAPPING:
+                mapped_key = mapped_key.replace(emb, EMBEDDING_DICT_MAPPING[emb])
+
+            # 2. Replace attention
+            if 'encoder' in each_key and 'self_attn' in each_key:
+                mapped_key = mapped_key.replace('self_attn', 'attention')
+            elif 'decoder' in each_key and 'self_attn' in each_key:
+                mapped_key = mapped_key.replace('self_attn', 'self_attention')
+            elif 'decoder' in each_key and 'encoder_attn' in each_key:
+                mapped_key = mapped_key.replace('encoder_attn', 'encoder_attention')
+
+            # 3. Replace multihead linear layers
+            #    fairseq sometimes chunks all three layers into one model weight
+            if 'in_proj_weight' in mapped_key or 'in_proj_bias' in mapped_key:
+                for weightorbias in {'weight', 'bias'}:
+                    attention_project_name = 'in_proj_{}'.format(weightorbias)
+                    if attention_project_name in mapped_key:
+                        weight = state_dict[each_key]
+                        size = int(weight.size(0) / 3)
+                        weights = weight.split(size, 0)
+                        # For Q, K, V in order
+                        return_dict[
+                            mapped_key.replace(
+                                attention_project_name, 'q_lin.{}'.format(weightorbias)
+                            )
+                        ] = weights[0]
+                        return_dict[
+                            mapped_key.replace(
+                                attention_project_name, 'k_lin.{}'.format(weightorbias)
+                            )
+                        ] = weights[1]
+                        return_dict[
+                            mapped_key.replace(
+                                attention_project_name, 'v_lin.{}'.format(weightorbias)
+                            )
+                        ] = weights[2]
+                continue
+            elif (
+                'v_proj' in mapped_key
+                or 'k_proj' in mapped_key
+                or 'q_proj' in mapped_key
+            ):
+                mapped_key = mapped_key.replace('v_proj', 'v_lin')
+                mapped_key = mapped_key.replace('q_proj', 'q_lin')
+                mapped_key = mapped_key.replace('k_proj', 'k_lin')
+
+            # 4. Replace FFN layers
+            mapped_key = mapped_key.replace('fc1', 'ffn.lin1')
```

Comment:
can we do these things in a loop with some set of tuples

Reply:
sure